app/models/reduction_fetcher.rb

```diff
@@ -35,9 +35,23 @@ def reduce_by_subject?
     @topic == :reduce_by_subject
   end
 
-  def where(query)
-    return @subject_reductions.where(query.except(:user_id)) if reduce_by_subject?
-    return @user_reductions.where(query.except(:subject_id)) if reduce_by_user?
+  def where(selector, in_place: false)
```

Comment:
Passing a boolean with a default value signals to me that this can be split into two methods, one for default, `!in_place` and `in_place` behaviour https://martinfowler.com/bliki/FlagArgument.html

One benefit is it removes the branching logic in the resulting method making it easier to understand as there is less logic.

Reply:
The branch still has to happen somewhere, though.  I'm happy to merge the `#where` and `#retrieve` methods if you think it will make the code clearer, but I don't see why the calling code should need to be aware of the implementation details--to the extent of invoking a separate method, at least--of how the reduction fetcher decides to locate the new records.

Reply:
I wouldn't merge the `#where` and `#retrieve` methods, but further break them apart. `#where` is pretty heavy and isn't very obvious/has a lot of branching logic. I think the Fowlerian approach being suggested would be to break it up into several more descriptively named methods for each of your conditions.

Reply:
#where has been split